mcnp2cad.cpp

```diff
@@ -785,24 +785,53 @@ entity_collection_t GeometryContext::defineCell(  CellCard& cell,  bool defineEm
     case CellCard::SURFNUM:
       {      
         int surface = token.second;
-        bool pos = true;
+        bool positive = true;
         if( surface < 0){
-          pos = false; surface = -surface;
+          positive = false; surface = -surface;
         }
         try{
           SurfaceVolume& surf = makeSurface( deck.lookup_surface_card( surface ) );
-          iBase_EntityHandle surf_handle = surf.define( pos, igm, world_size );
+          iBase_EntityHandle surf_handle = surf.define( positive, igm, world_size );
           stack.push_back(surf_handle);
         }
         catch(std::runtime_error& e) { std::cerr << e.what() << std::endl; }
       }
       break;
     case CellCard::MBODYFACET:
       {
-        int cellnum = std::abs(token.second) / 10;
-        int facet = std::abs(token.second) - (cellnum*10);
-        std::cerr << "Attempting to define facet " << facet << " of cell " << cellnum << std::endl;
-        throw std::runtime_error( "Macrobody facets are not yet supported by mcnp2cad." );
+        int identifier = -std::abs( token.second );
+        int surfacenum = -identifier / 10;
+        int facet = -identifier - ( surfacenum * 10 );
+
+        try{
+          makeSurface( deck.lookup_surface_card( identifier ), NULL, facet);
+          SurfaceVolume& surf = makeSurface( deck.lookup_surface_card( identifier ) );
+          const std::string& mnemonic = deck.lookup_surface_card( identifier )->getMnemonic();
+          bool positive = true;
+          if( mnemonic == "rcc" || mnemonic == "rec" ){
+            if( ( token.second < 0 ) ^ ( facet == 3 ) ){
+              positive = false;
+            }
+          }
+          else if( mnemonic == "box" || mnemonic == "rpp" ){
+            if( ( token.second < 0 ) ^ ( facet == 2 || facet == 4 || facet == 6 ) ){
+              positive = false;
+            }
+          }
+          else if( mnemonic == "hex" || mnemonic == "rhp" ){
+            if( ( token.second < 0 ) ^ ( facet == 8 ) ){
```

Comment:
I think it is because with the hex and rhp sides, it defines all of them as planes at the end of a vector, while for all the other planes created perpendicular to vectors a plane is created at the beginning and at the end.  The one at the beginning then needs to be flipped.
Either way, this works with every test I've run.